### PR TITLE
speed test algorithm implementation for redundant streaming

### PR DESCRIPTION
**Pull request by obinnaokechukwu**
Merging speed test changes with dual-ingest-q1 branch.

### CODE REVIEW

File: UI/window-basic-auto-config-test.cpp [hunk 1]
```diff
@@ -384,47 +482,89 @@ void AutoConfigTestPage::TestBandwidthThread()
...
-			server.bitrate = (int)bitrate * 70 / 100;
+		int avg_bitrate = (int)bitrate / urls.size();
+
+		if (frames_dropped ||
+		    (int)avg_bitrate < (wiz->startingBitrate * 75 / 100)) {
+			server.bitrate = (int)avg_bitrate * 70 / 100;
 		} else {
 			server.bitrate = wiz->startingBitrate;
+			server.preferred = wiz->allowRedundantStreams &&
+					   urls.size() > 1;
```
Comment (obinnaokechukwu) [inline UI/window-basic-auto-config-test.cpp:558]: Flag is used to choose dual ingestion.

File: UI/window-basic-auto-config-test.cpp [hunk 2]
```diff
@@ -439,16 +579,19 @@ void AutoConfigTestPage::TestBandwidthThread()
...
 	string bestServerName;
+	bool isBestPreferred = false;
 
 	for (auto &server : servers) {
 		bool close = abs(server.bitrate - bestBitrate) < 400;
 
 		if ((!close && server.bitrate > bestBitrate) ||
-		    (close && server.ms < bestMS)) {
+		    (close && server.ms < bestMS) ||
+		    (server.preferred && !isBestPreferred)) {
```
Comment (obinnaokechukwu) [inline UI/window-basic-auto-config-test.cpp:589]: preferred server is given precedence over non-preferred servers.

File: UI/window-basic-auto-config-test.cpp [hunk 3]
```diff
@@ -384,47 +482,89 @@ void AutoConfigTestPage::TestBandwidthThread()
...
+
+		if (frames_dropped ||
+		    (int)avg_bitrate < (wiz->startingBitrate * 75 / 100)) {
+			server.bitrate = (int)avg_bitrate * 70 / 100;
+			server.preferred = false;
 		} else {
 			server.bitrate = wiz->startingBitrate;
+			server.preferred = wiz->allowRedundantStreams &&
+					   urls.size() > 1;
+		}
```
Comment (obinnaokechukwu) [inline UI/window-basic-auto-config-test.cpp:560]: `server.preferred` flag is used to choose dual ingestion.

File: UI/window-basic-auto-config-test.cpp [hunk 4]
```diff
@@ -262,89 +277,132 @@ void AutoConfigTestPage::TestBandwidthThread()
...
 	/* -----------------------------------*/
 	/* connect signals                    */
 
-	auto on_started = [&]() {
-		unique_lock<mutex> lock(m);
-		connected = true;
-		stopped = false;
-		cv.notify_one();
-	};
+	function<void(struct OutputWrapper &)> on_started =
```
Comment (ushadow) [inline UI/window-basic-auto-config-test.cpp:334]: Is there a reason why auto can't be used here?

File: UI/window-basic-auto-config.hpp [hunk 5]
```diff
@@ -245,6 +246,7 @@ class AutoConfigTestPage : public QWizardPage {
 		std::string address;
 		int bitrate = 0;
 		int ms = -1;
+		bool preferred = false;
```
Comment (ushadow) [inline UI/window-basic-auto-config.hpp:249]: Maybe add a comment for the new field?

File: UI/window-basic-auto-config-test.cpp (hunk 4 above)
Comment (obinnaokechukwu) [reply UI/window-basic-auto-config-test.cpp:334]: The function is eventually put into a struct which needs exact type definition.

**Pull request merged**